Nested patterns in mlang

elegios opened the pull request:
This PR changes patterns in `sem`antic functions in mlang to be the normal patterns present in mexpr. For pre-existing semantic functions this should have no externally visible change, but new ones gain additional powers.

When calling a semantic function the most specific pattern that matches the argument will dictate the arm that gets executed. Note that this means that the order of the arms in the code does not matter, ordering is determined solely by specificity. Specificity is defined as subset on the set of values a pattern can match. For example, `(true, true)` is more specific than `(true, _)` since the set of matchable values is a subset.

This also creates a new failure case for merging semantic functions: if there are two patterns that overlap, but neither is more specific than the other, then the composition fails. At present there is no way to handle this failure (e.g. by manually specifying which branch to take, or adding a new pattern that matches the intersection of the overlapping patterns) but we have thoughts on how to do this in the future.

The code that performs this analysis does not have great time complexity (this is expected, checking for redundancy in pattern matches is NP-complete). However, cases we run into in practice should still be fast, since the patterns are small, but please report slowdowns.

Additionally, and unrelated, it also removes `unit` from the initial environment of values, and `PatUnit` from patterns.

elegios committed:
Remove 'unit'

elegios committed:
Added nested patterns to mlang + specificity analysis of patterns

elegios committed:
Fix compare for the sets in patterns.ml + test nested patterns in mlang

elegios committed:
Bugfix for intersection between records and record negations + more tests + better error message

EliasC commented on src/boot/msg.ml:
```diff
@@ -41,6 +41,11 @@ type message = id * severity * info * arguments
+let eq_info a b = match a, b with
+  | NoInfo, NoInfo -> true
+  | Info(f1, l11, c11, l21, c21), Info(f2, l12, c12, l22, c22) ->
+     l11 = l12 && c11 = c12 && l21 == l22 && c21 == c22 && f1 =. f2
```
Is there a motivation for comparing the starting coordinates with `=` and the latter with `==`?

elegios replied:
No, it's a typo. Thanks.

EliasC commented on src/boot/mlang.ml:
```diff
@@ -8,140 +8,171 @@
+  | (TopLet _ | TopRecLet _ | TopCon _) as top -> (prev_langs, top)
+  | TopLang(Lang(info, name, includes, _) as lang) ->
+     let self_data = compute_lang_data lang in
+     let included_data = List.map (lookup_lang info prev_langs) includes in
```
This isn't really a list of data, but a list of langs, right?

elegios replied:
It's a `lang_data list`. I don't particularly like the name of this type, but I didn't come up with anything better. It's supposed to be a representation of the results of the calculations I need for a language for my analyses here, so essentially caching (though I prefer to think of it as a more accurate representation, e.g., the cases are put in a DAG instead of just in a list, since the order in the list doesn't actually matter).

EliasC reviewed (approved):
This looks OK to me. I did not get into the nitty gritty details of `patterns.ml`.

elegios committed:
Be consistent in equality operator used in eq_info

elegios committed:
Fix bug in intersection between edge patterns + more tests

If we are taking the intersection between two edge patterns, where
one pattern has a longer prefix than the other, and the other has
a longer postfix than the other, then the intersection should contain
some exact length lists that are shorter than (max pre1 pre2) +
(max post1 post2), namely those where the extraneous patterns overlap.
For example, [!1] ++ _ & _ ++ [!2] is the same as
[!1] ++ _ ++ [!2] | [!(1 | 2)], where the second part of the union would
not show up in the old implementation.
